engine/execution/ingestion/engine_test.go

```diff
@@ -276,111 +271,191 @@ func (ctx *testingContext) assertSuccessfulBlockComputation(executableBlock *ent
+
+		wg.Add(1) // wait for block A to be executed
+		err := ctx.engine.handleBlock(context.Background(), blockA.Block)
+		require.NoError(t, err)
 
-	execState.AssertExpectations(t)
+		unittest.AssertReturnsBefore(t, wg.Wait, 5*time.Second)
+
+		_, more := <-ctx.engine.Done() //wait for all the blocks to be processed
+		require.False(t, more)
+
+		_, ok := commits[blockA.ID()]
+		require.True(t, ok)
+	})
 }
 
-func TestBlockOutOfOrder(t *testing.T) {
+func logBlocks(blocks map[string]*entity.ExecutableBlock) {
+	log := unittest.Logger()
+	for name, b := range blocks {
+		log.Debug().Msgf("creating blocks for testing, block %v's ID:%v", name, b.ID())
+	}
+}
 
+func TestExecuteBlockInOrder(t *testing.T) {
 	runWithEngine(t, func(ctx testingContext) {
 
-		executableBlockA := unittest.ExecutableBlockFixture(nil)
-		executableBlockB := unittest.ExecutableBlockFixtureWithParent(nil, executableBlockA.Block.Header)
-		executableBlockC := unittest.ExecutableBlockFixtureWithParent(nil, executableBlockA.Block.Header)
-		executableBlockD := unittest.ExecutableBlockFixtureWithParent(nil, executableBlockC.Block.Header)
-		executableBlockA.StartState = unittest.StateCommitmentFixture()
+		// create blocks with the following relations
+		// A <- B
+		// A <- C <- D
+		blocks := make(map[string]*entity.ExecutableBlock)
+		blocks["A"] = unittest.ExecutableBlockFixture(nil)
+		blocks["A"].StartState = unittest.StateCommitmentFixture()
 
-		// blocks has no collections, so state is essentially the same
-		executableBlockC.StartState = executableBlockA.StartState
-		executableBlockB.StartState = executableBlockA.StartState
-		executableBlockD.StartState = executableBlockC.StartState
+		blocks["B"] = unittest.ExecutableBlockFixtureWithParent(nil, blocks["A"].Block.Header)
+		blocks["C"] = unittest.ExecutableBlockFixtureWithParent(nil, blocks["B"].Block.Header)
+		blocks["D"] = unittest.ExecutableBlockFixtureWithParent(nil, blocks["C"].Block.Header)
 
-		/* Artists recreation of the blocks structure:
+		// log the blocks, so that we can link the block ID in the log with the blocks in tests
+		logBlocks(blocks)
 
-		  b
-		   \
-		    a
-		   /
-		d-c
+		// none of the blocks has any collection, so state is essentially the same
+		blocks["C"].StartState = blocks["A"].StartState
+		blocks["B"].StartState = blocks["A"].StartState
+		blocks["D"].StartState = blocks["C"].StartState
 
-		*/
+		commits := make(map[flow.Identifier]flow.StateCommitment)
+		commits[blocks["A"].Block.Header.ParentID] = blocks["A"].StartState
 
-		ctx.blocks.EXPECT().Store(gomock.Eq(executableBlockA.Block))
-		ctx.blocks.EXPECT().Store(gomock.Eq(executableBlockB.Block))
-		ctx.blocks.EXPECT().Store(gomock.Eq(executableBlockC.Block))
-		ctx.blocks.EXPECT().Store(gomock.Eq(executableBlockD.Block))
+		wg := sync.WaitGroup{}
+		ctx.mockStateCommitsWithMap(commits, func(blockID flow.Identifier, commit flow.StateCommitment) {
+			wg.Done()
+		})
```

Comment:
Why do we need the wait group here? Is waiting on `engine.Done()` not sufficient?

Reply:
yeah, `engine.Done` is not suficient.

For instance:
```
A <- B
    ^- C <- D
```

When I send block A-D to the ingestion engine, it will only execute block A, because B and C has to wait for A's result. 
When I call `Done`, it's possible that B and C has started, so `<-Done` will wait until both B and C are finished. However, `D` won't be executed, because the engine has stopped. 

So I have to explicitly say "wait until A, B, C, D all have been executed"

Reply:
👌 Gotcha, makes sense